statsmodels/gam/gam.py

```diff
@@ -0,0 +1,767 @@
+            res_ = self._fit_pirls(start_params=history['params'][-1],
+                                   alpha=a)
+            history['alpha'].append(a)
+            history['params'].append(np.asarray(res_.params))
+            return getattr(res_, criterion)
+
+        if method == 'nm':
+            kwds = dict(full_output=True, maxiter=1000, maxfun=2000)
+            kwds.update(fit_kwds)
+            fit_res = optimize.fmin(fun, start_params, **kwds)
+            opt = fit_res[0]
+        elif method == 'basinhopping':
+            kwds = dict(minimizer_kwargs={'method': 'Nelder-Mead',
+                        'options': {'maxiter': 100, 'maxfev': 500}},
+                        niter=10)
+            kwds.update(fit_kwds)
+            fit_res = optimize.basinhopping(fun, start_params, **kwds)
+            opt = fit_res.x
+        elif method == 'minimize':
+            fit_res = optimize.minimize(fun, start_params, **fit_kwds)
+            opt = fit_res.x
+        else:
+            raise ValueError('method not recognized')
+
+        del history['params'][0]  # remove the model start_params
+
+        alpha = np.exp(opt)
+        return alpha, fit_res, history
+
+
+class LogitGam(PenalizedMixin, Logit):
+    """Generalized Additive model for discrete Logit
+
+    This subclasses discrete_model Logit.
+
+    Warning: not all inherited methods might take correctly account of the
+    penalization
+
+    not verified yet.
+
+    """
+    def __init__(self, endog, smoother, alpha, *args, **kwargs):
+        import collections
+        if not isinstance(alpha, collections.Iterable):
+            alpha = np.array([alpha] * len(smoother.smoothers))
+
+        self.smoother = smoother
+        self.alpha = alpha
+        self.pen_weight = 1  # TODO: pen weight should not be defined here!!
+        penal = MultivariateGamPenalty(smoother, alpha=alpha)
+
+        super(LogitGam, self).__init__(endog, smoother.basis, penal=penal,
+                                       *args, **kwargs)
+
+
+def penalized_wls(endog, exog, penalty_matrix, weights):
+    """weighted least squares with quadratic penalty
+
+    Parameters
+    ----------
+    endog : ndarray
+        response or endogenous variable
+    exog : ndarray
+        design matrix, matrix of exogenous or explanatory variables
+    penalty_matrix : ndarray, 2-Dim square
+        penality matrix for quadratic penalization. Note, the penalty_matrix
+        is multiplied by two to match non-pirls fitting methods.
+    weights : ndarray
+        weights for WLS
+
+    Returns
+    -------
+    results : Results instance of WLS
+    """
+    y, x, s = endog, exog, penalty_matrix
+    # TODO: I don't understand why I need 2 * s
+    aug_y, aug_x, aug_weights = make_augmented_matrix(y, x, 2 * s, weights)
+    wls_results = lm.WLS(aug_y, aug_x, aug_weights).fit()
+    # MinimalWLS does not return normalized_cov_params
+    # wls_results = reg_tools._MinimalWLS(aug_y, aug_x, aug_weights).fit()
```

Comment:
this comment could use clarification.  is it saying "we would use MinimalWLS, but [...] 
 If we did, this would be how we call it [...]"?

Reply:
nice, thanks